hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/GaugeImpl.java

```diff
@@ -0,0 +1,150 @@
+    volatile S source;
+
+    private final ILogger logger;
+    private final String name;
+
+    public GaugeImpl(String name, ILogger logger) {
+        this.name = name;
+        this.logger = logger;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public void render(StringBuilder sb) {
+        checkNotNull(sb, "sb can't be null");
+
+        sb.append(name).append('=');
+
+        Object input = this.input;
+        S source = this.source;
+
+        if (input == null || source == null) {
+            sb.append("NA");
+            return;
+        }
+
+        try {
+            if (input instanceof LongProbe) {
+                LongProbe function = (LongProbe) input;
+                sb.append(function.get(source));
+            } else if (input instanceof DoubleProbe) {
+                DoubleProbe function = (DoubleProbe) input;
+                sb.append(function.get(source));
+            } else if (input instanceof AccessibleObjectGauge) {
+                AccessibleObjectGauge accessibleObjectInput = (AccessibleObjectGauge) input;
+                if (accessibleObjectInput.isDouble()) {
+                    sb.append(accessibleObjectInput.getDouble(source));
+                } else {
+                    sb.append(accessibleObjectInput.getLong(source));
+                }
+            }
+        } catch (Exception e) {
+            logger.warning("Failed to update metric:" + name, e);
+            sb.append("NA");
+        }
+    }
+
+    @Override
+    public long readLong() {
+        Object input = this.input;
+        S source = this.source;
+        long result = 0;
+
+        if (input == null || source == null) {
+            return result;
+        }
+
+        try {
+            if (input instanceof AccessibleObjectGauge) {
+                AccessibleObjectGauge accessibleObjectInput = (AccessibleObjectGauge) input;
+                if (accessibleObjectInput.isDouble()) {
+                    double doubleResult = accessibleObjectInput.getDouble(source);
+                    result = Math.round(doubleResult);
+                } else {
+                    result = accessibleObjectInput.getLong(source);
+                }
+            } else if (input instanceof LongProbe) {
+                LongProbe<S> longInput = (LongProbe<S>) input;
+                result = longInput.get(source);
+            } else if (input instanceof DoubleProbe) {
+                DoubleProbe<S> doubleInput = (DoubleProbe<S>) input;
+                double doubleResult = doubleInput.get(source);
+                result = Math.round(doubleResult);
+            }
+            return result;
+        } catch (Exception e) {
+            logger.warning("Failed to update metric:" + name, e);
+            return result;
+        }
+    }
+
+    @Override
+    public double readDouble() {
+        Object input = this.input;
+        S source = this.source;
+        double result = 0;
```

Comment:
No. There can be 2 kind of probes. Long probes and double probes. CollectionSizeProbe is a long probe, MapSizeProbe is also a long probe. If a field or method has a double return value, it can be a DoubleFieldProbe or DoubleMethodProbe. You can wrap these probes with corresponding LongProbeReader or DoubleProbeReader objects and there wont be a duplication. In the current code, there are if-else / switch statements everywhere.